substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/resources/NativeImageResourceFileSystem.java

```diff
@@ -0,0 +1,1217 @@
+        beginWrite();
+        try {
+            IndexNode rootIndex = inodes.get(lookupKey.as(ROOT_PATH));
+            if (rootIndex == null) {
+                rootIndex = new IndexNode(ROOT_PATH, true);
+            } else {
+                inodes.remove(rootIndex);
+            }
+            IndexNode[] nodes = inodes.keySet().toArray(new IndexNode[0]);
+            inodes.put(rootIndex, rootIndex);
+            ParentLookup lookup = new ParentLookup();
+            for (IndexNode controlNode : nodes) {
+                IndexNode parent;
+                IndexNode node = controlNode;
+                while (true) {
+                    int off = getParentOff(node.name);
+                    if (off <= 1) {    // Parent is root.
+                        node.sibling = rootIndex.child;
+                        rootIndex.child = node;
+                        break;
+                    }
+                    lookup = lookup.as(node.name, off);
+                    if (inodes.containsKey(lookup)) {
+                        parent = inodes.get(lookup);
+                        node.sibling = parent.child;
+                        parent.child = node;
+                        break;
+                    }
+                    // Add new pseudo directory entry.
+                    parent = new IndexNode(Arrays.copyOf(node.name, off), true);
+                    inodes.put(parent, parent);
+                    node.sibling = parent.child;
+                    parent.child = node;
+                    node = parent;
+                }
+            }
+        } finally {
+            endWrite();
+        }
+    }
+
+    private InputStream getInputStream(Entry e) throws IOException {
+        InputStream eis = null;
+        if (e.type == Entry.NEW || e.type == Entry.COPY) {
+            byte[] bytes = e.getBytes(true);
+            if (bytes != null) {
+                eis = new ByteArrayInputStream(bytes);
+            } else {
+                if (e.file != null) {
+                    eis = Files.newInputStream(e.file);
+                } else {
+                    throw new NativeImageResourceFileSystemException("Entry data is missing!");
+                }
+            }
+        } else {
+            if (e.type == Entry.FILE_CH) {
+                eis = Files.newInputStream(e.file);
+                return eis;
+            }
+        }
+        streams.add(eis);
+        return eis;
+    }
+
+    private OutputStream getOutputStream(Entry e) throws IOException {
+        e.getBytes(false);
+        if (e.lastModifiedTime == -1) {
+            e.lastModifiedTime = System.currentTimeMillis();
+        }
+        OutputStream os;
+        if (useTempFile) {
+            e.file = getTempPathForEntry(null);
+            os = Files.newOutputStream(e.file, WRITE);
+        } else {
+            os = new ByteArrayOutputStream((e.size > 0) ? e.size : 8192);
+        }
+        return new EntryOutputStream(e, os);
```

Comment:
If we keep everything in memory, would it make sense to also close the output streams when closing the file system?

Reply:
Agree. I overlooked it. We are keeping all information about opened input streams, so we should do the same thing for output streams.